core/utils/hashset/hashset.c

```diff
@@ -0,0 +1,153 @@
+/* -*- Mode: C; tab-width: 4; c-basic-offset: 4; indent-tabs-mode: nil -*- */
+/*
+ *     Copyright 2012 Couchbase, Inc.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ * 
+ * Modified in 2022 by Edward A. Lee (eal@berkeley.edu) so that stored items are
+ * consistently of type void*. Note that the void* value 1 is used to mark a deleted
+ * item and therefore cannot be stored in the hashset.
+ * Also, used float rather than double for determining whether to rehash.
+ */
+
+#include "hashset/hashset.h"
+#include <assert.h>
+
+static const unsigned int prime_1 = 73;
+static const unsigned int prime_2 = 5009;
+
+hashset_t hashset_create(unsigned short nbits) {
+    hashset_t set = calloc(1, sizeof(struct hashset_st));
+
+    if (set == NULL) {
+        return NULL;
+    }
+    set->nbits = nbits;
+    set->capacity = (size_t)(1 << set->nbits);
+    set->mask = set->capacity - 1;
+    set->items = (void**)calloc(set->capacity, sizeof(void*));
+    if (set->items == NULL) {
+        hashset_destroy(set);
+        return NULL;
+    }
+    set->nitems = 0;
+    set->n_deleted_items = 0;
+    return set;
+}
+
+size_t hashset_num_items(hashset_t set) {
+    return set->nitems;
+}
+
+void hashset_destroy(hashset_t set) {
+    if (set) {
+        free(set->items);
+    }
+    free(set);
+}
+
+// FIXME
+#include <stdio.h>
+
+static int hashset_add_member(hashset_t set, void *item) {
+    size_t ii;
+
+    if (item == 0 || item == (void*)1) {
+        return -1;
+    }
+    ii = set->mask & (prime_1 * (size_t)item);
+
+    // Search chain of possible locations and stop when slot is empty.
+    // Chain of possibilities always ends with an empty (0) even
+    // if some items in the chain have been deleted (1). If a
+    // deleted slot is found along the way, remember it to use it.
+    int available = -1;
+    while (set->items[ii] != 0) {
+        if (set->items[ii] == item) {
+            return 0;
+        } else {
+            if (set->items[ii] == (void*)1 && available < 0) {
+                // Slot is available from deletion.
+                available = (int)ii;
```

Comment:
Just to understand it right, I have a question please.
Why don't we add the entry here at this available slot and return, since we've already found our preferred slot?

Reply:
Because the item being added may still be in the hashset. If there was previously a collision of hashes, where three items were added to the set, we will get a chain like this:
```
A -> B -> C -> 0
```
If `B` later gets removed, we will have:
```
A -> 1 -> C -> 0
```
If we then try to add `C`, we cannot just put in the `1` slot because then we will have two `C`s in the hashset.

Reply:
thanks Professor. This is super necessary to avoid duplication, and we remember the first deleted slot because we want to add this one as early as possible in the chain.